crates/circuit/src/parameter_expression.rs

```diff
@@ -333,173 +1834,565 @@ impl ParameterExpression {
     // ====================================
     pub fn __eq__(&self, rhs: &Bound<PyAny>) -> bool {
         match _extract_value(rhs) {
-            Some(rhs) => match rhs.expr {
-                SymbolExpr::Value(v) => match self.expr.eval(true) {
-                    Some(e) => e == v,
-                    None => false,
-                },
-                _ => self.expr == rhs.expr,
-            },
+            Some(rhs) => self.compare_eq(&rhs, true),
             None => false,
         }
     }
     pub fn __ne__(&self, rhs: &Bound<PyAny>) -> bool {
+        !self.__eq__(rhs)
+    }
+
+    pub fn __lt__(&self, rhs: &Bound<PyAny>) -> bool {
         match _extract_value(rhs) {
-            Some(rhs) => match rhs.expr {
-                SymbolExpr::Value(v) => match self.expr.eval(true) {
-                    Some(e) => e != v,
-                    None => true,
-                },
-                _ => self.expr != rhs.expr,
-            },
-            None => true,
+            Some(rhs) => self.expr < rhs.expr,
+            None => false,
         }
     }
-    pub fn __neg__(&self) -> Self {
-        Self { expr: -&self.expr }
+    pub fn __gt__(&self, rhs: &Bound<PyAny>) -> bool {
+        match _extract_value(rhs) {
+            Some(rhs) => self.expr > rhs.expr,
+            None => false,
+        }
+    }
+
+    // unary operators
+    pub fn __neg__(&self) -> ParameterExpression {
```

Comment:
I'm a bit curious why these are all defined twice; once with a leading underscore and then the public method that users are supposed to call (e.g. `__neg__` calls `_neg` or `sin` calls `_sin`). Is there a specific reason for this? 🙂